lib/ansible/plugins/action/copy.py

```diff
@@ -208,6 +208,13 @@ class ActionModule(ActionBase):
 
     TRANSFERS_FILES = True
 
+    def _ensure_invocation(self, result):
+        if result.get('failed', False) and self._play_context.no_log:
+            result['invocation'] = "CENSORED: no_log is set"
+        elif not result.has_key("invocation"):
+            result["invocation"] = self._task.args
+        return result
```

Comment:
I don't think the logic is right here.

* For failed==False, I think that the module side handles the play_context's no_log?  (IIRC, we pass that into the module as _ansible_no_log?).
* It feels like we should only be calling this for failed==True.  failed=False should be handled entirely module-side.  So I'd do the failed=True|False check in the calling code and only call this if we've failed.
* The check to add invocation needs to always depend on play_context.no_log being False.  If there's any case where no_log is True and we still add the invocation we'll be violating the security guarantees of no_log.

Reply:
I'd also make this a toplevel function rather than a method because self is unused inside of the function.

Reply:
I don't understand what you mean by "module side" in this scenario, the module never gets called in the execution path reported in the issue ticket this is aiming to fix (which is ultimately the problem).

Reply:
You mentioned `self` isn't used but I'm referencing `self._task.args`, do you mean the args should just be passed in?

Reply:
@maxamillion regarding self.  Oops, I didn't see that when I made that comment.  I was wrong.

Ah, I think I was getting confused because the security issue we recently fixed is also about no_log but it is only about no_log when failed.

however, the logic still seems wrong.  It shouldn't it be something like:
```
if 'invocation' not in result:
    if self._play_context.no_log:
        result['invocation'] = "CENSORED: no_log is set"
    else:
        result["invocation"] = self._task.args

Reply:
@abadger yes, that absolutely makes more sense. 👍